sql/core/src/main/scala/org/apache/spark/sql/execution/command/tables.scala

```diff
@@ -734,15 +736,290 @@ case class DescribeTableCommand(
     append(buffer, "# Detailed Partition Information", "", "")
     append(buffer, "Database", table.database, "")
     append(buffer, "Table", tableIdentifier.table, "")
-    partition.toLinkedHashMap.foreach(s => append(buffer, s._1, s._2, ""))
+    toLinkedHashMap(partition.toJsonLinkedHashMap)
+      .foreach(s => append(buffer, s._1, s._2, ""))
     append(buffer, "", "", "")
     append(buffer, "# Storage Information", "", "")
     table.bucketSpec match {
       case Some(spec) =>
-        spec.toLinkedHashMap.foreach(s => append(buffer, s._1, s._2, ""))
+        toLinkedHashMap(spec.toJsonLinkedHashMap)
+          .foreach(s => append(buffer, s._1, s._2, ""))
       case _ =>
     }
-    table.storage.toLinkedHashMap.foreach(s => append(buffer, s._1, s._2, ""))
+    toLinkedHashMap(table.storage.toJsonLinkedHashMap)
+      .foreach(s => append(buffer, s._1, s._2, ""))
+  }
+}
+
+/**
+ * Command that looks like
+ * {{{
+ *   DESCRIBE [EXTENDED|FORMATTED] table_name partitionSpec? [AS JSON];
+ * }}}
+ */
+case class DescribeTableJsonCommand(
+   table: TableIdentifier,
+   partitionSpec: TablePartitionSpec,
+   isExtended: Boolean)
+  extends LeafRunnableCommand {
+  override val output = DescribeCommandSchema.describeJsonTableAttributes()
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    val result = new ArrayBuffer[Row]
+    val catalog = sparkSession.sessionState.catalog
+
+    if (catalog.isTempView(table)) {
+      if (partitionSpec.nonEmpty) {
+        throw QueryCompilationErrors.descPartitionNotAllowedOnTempView(table.identifier)
+      }
+      val schema = catalog.getTempViewOrPermanentTableMetadata(table).schema
+      describeColsJson(schema, result, header = false)
+    } else {
+      val metadata = catalog.getTableRawMetadata(table)
+      val schema = if (metadata.schema.isEmpty) {
+        // In older versions of Spark,
+        // the table schema can be empty and should be inferred at runtime.
+        sparkSession.table(metadata.identifier).schema
+      } else {
+        metadata.schema
+      }
+
+      addKeyValueToJson(result, "table_name", JString(metadata.identifier.table))
+
+      table.catalog.foreach(catalog => addKeyValueToJson(result, "catalog_name", JString(catalog)))
+      table.database.foreach { db =>
+        addKeyValueToJson(result, "namespace", JArray(List(JString(db))))
+        addKeyValueToJson(result, "schema_name", JString(db))
+      }
+
+      describeColsJson(schema, result, header = false)
+
+      describeClusteringInfoJson(metadata, result)
+
+      if (partitionSpec.nonEmpty) {
+        // Outputs the partition-specific info for the DDL command:
+        // "DESCRIBE [EXTENDED|FORMATTED] table_name PARTITION (partitionVal*)"
+        describePartitionInfoJson(sparkSession, catalog, metadata, result)
+      } else {
+        describeFormattedTableInfoJson(metadata, result)
+      }
+    }
+
+    result.toSeq
```

Comment:
If we expect users to parse the JSON string of this AS JSON command, we must return only one row containing the full JSON string.

Reply:
And I'm wondering if it's simply to construct a `Map[String, JValue]` during this command, and finally call `pretty(render(JObject(jsonMap)))` to generate the full pretty JSON string?

Reply:
makes sense, I updated to construct a map and then convert the final map to JSON string
just noting that I used `compact(...)` instead of `pretty(...)` because it may be better for parsing purposes